apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts

```diff
@@ -0,0 +1,201 @@
+import { captureException } from '@sentry/nextjs'
+
+import Bottleneck from 'bottleneck'
+import { Address, Hex } from 'viem'
+import { WallchainKeys } from 'config/wallchain'
+import { useSwapCallArguments } from './useSwapCallArguments'
+
+interface SwapCall {
+  address: Address
+  calldata: Hex
+  value: Hex
+}
+interface WallchainSwapCall {
+  getCall: () => Promise<SwapCall>
+}
+
+export type WallchainStatus = 'found' | 'pending' | 'not-found'
+
+const limiter = new Bottleneck({
+  maxConcurrent: 1, // only allow one request at a time
+  minTime: 250, // add 250ms of spacing between requests
+  highWater: 1, // only queue 1 request at a time, newer request will drop older
+})
+
+const overrideAddresses = {
+  56: '0x6346e0a39e2fBbc133e4ce8390ab567108e62aEe',
+}
+
+const loadData = async (account: string, sdk: WallchainSDK, swapCalls: SwapCall[]) => {
+  if (await sdk.supportsChain()) {
+    const resp = await sdk.checkForMEV({
+      from: account,
+      to: swapCalls[0].address,
+      value: swapCalls[0].value,
+      data: swapCalls[0].calldata,
+    })
+
+    if (resp.MEVFound) {
+      const approvalFor = await sdk.getSpenderForAllowance()
+      return ['found', approvalFor, resp.masterInput]
+    }
+  }
+  return ['not-found', undefined, undefined]
+}
+const wrappedLoadData = limiter.wrap(loadData)
+
+const extractAddressFromCurrency = (currency: Currency): `0x${string}` => {
+  return currency.isNative ? '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee' : (currency as Token).address
+}
+
+const extractTokensFromTrade = (trade: SmartRouterTrade<TradeType> | undefined | null) => {
+  const inputCurrency = trade?.inputAmount?.currency
+  const outputCurrency = trade?.outputAmount?.currency
+  const srcToken = inputCurrency ? extractAddressFromCurrency(inputCurrency) : false
+  const dstToken = outputCurrency ? extractAddressFromCurrency(outputCurrency) : false
+
+  return [srcToken, dstToken]
+}
+
+export function useWallchainApi(
+  trade?: SmartRouterTrade<TradeType>,
+  deadline?: bigint,
+  feeOptions?: FeeOptions,
+): [WallchainStatus, string | undefined, string | undefined] {
+  const [approvalAddress, setApprovalAddress] = useState<undefined | string>(undefined)
+  const [status, setStatus] = useState<WallchainStatus>('pending')
+  const [masterInput, setMasterInput] = useState<undefined | string>(undefined)
+  const { data: walletClient } = useWalletClient()
+  const { account } = useAccountActiveChain()
+  const [allowedSlippageRaw] = useUserSlippage() || [INITIAL_ALLOWED_SLIPPAGE]
+  const allowedSlippage = useMemo(() => basisPointsToPercent(allowedSlippageRaw), [allowedSlippageRaw])
+
+  const swapCalls = useSwapCallArguments(trade, allowedSlippage, account, deadline, feeOptions)
+
+  useEffect(() => {
+    if (!walletClient) {
+      return
+    }
+    const sdk = new WallchainSDK({
+      keys: WallchainKeys,
+      provider: walletClient.transport as TOptions['provider'],
+      overrideAddresses,
+    })
```

Comment:
Let's move the sdk initialisation inside the condition, would be great to use lazy import `await import('@wallchain/sdk')`